File: runtime/go/units/wasi/fd_prestat_get.go
// requires: memory/init
func (w *WASI) wasi_fd_prestat_get(fd, outPtr uint32) uint32 {
    d, ok := w.fds[fd].(*wasiDir)
    if !ok || d.preopenName == nil {
        return wasiBadf
    }
    // prestat: tag (u8, 0 = dir) + 3 pad + pr_name_len (u32).
    buf := make([]byte, 8)
    binary.LittleEndian.PutUint32(buf[4:8], uint32(len(d.preopenName)))
    w.memory.init(uint64(outPtr), buf, 0, 8)
    return wasiOk
}

File: runtime/python/units/wasi/fd_prestat_get.py
# requires: memory/init
def wasi_fd_prestat_get(self, fd, out_ptr):
    entry = self.fds.get(fd)
    if not isinstance(entry, self.WasiDir) or entry.preopen_name is None:
        return self.ERRNO_BADF
    # prestat: tag (u8, 0 = dir) + 3 pad + pr_name_len (u32).
    self.memory.init(out_ptr, struct.pack("<BxxxI", 0, len(entry.preopen_name)), 0, 8)
    return self.ERRNO_SUCCESS
